src/main/java/com/bloxbean/cardano/client/backend/api/helper/impl/FeeCalculationServiceImpl.java

```diff
@@ -47,10 +56,19 @@ public BigInteger calculateFee(PaymentTransaction paymentTransaction, Transactio
     public BigInteger calculateFee(PaymentTransaction paymentTransaction, TransactionDetailsParams detailsParams,
                                    Metadata metadata, ProtocolParams protocolParams) throws CborSerializationException, AddressExcepion, ApiException {
 
-        if(paymentTransaction.getFee() == null || paymentTransaction.getFee().compareTo(BigInteger.valueOf(170000)) == -1) //Just a dummy fee
-            paymentTransaction.setFee(new BigInteger("170000")); //Set a min fee just for calcuation purpose if not set
-        //Build transaction
-        String txnCBORHash = transactionHelperService.createSignedTransaction(Arrays.asList(paymentTransaction), detailsParams, metadata);
+        PaymentTransaction clonePaymentTransaction = paymentTransaction.toBuilder().build();
+        if(clonePaymentTransaction.getFee() == null || clonePaymentTransaction.getFee().compareTo(DUMMY_FEE) == -1) //Just a dummy fee
+            clonePaymentTransaction.setFee(DUMMY_FEE); //Set a min fee just for calcuation purpose if not set
+
+        String txnCBORHash;
+        try {
+            //Build transaction
+            txnCBORHash = transactionHelperService.createSignedTransaction(Arrays.asList(clonePaymentTransaction), detailsParams, metadata);
+        } catch (InsufficientBalanceException e) {
+            clonePaymentTransaction.setFee(MIN_DUMMY_FEE);
+            clonePaymentTransaction.setAmount(clonePaymentTransaction.getAmount().subtract(MIN_DUMMY_FEE));
```

Comment:
Is there no check required to make sure tx amount is lovelace?

Reply:
Right.  I have added a condition to check if unit = lovelace then retry. If not lovelace, then throw original exception.
Thanks.